OPHD/Things/Structures/Structure.cpp

```diff
@@ -60,6 +60,54 @@ const std::string& structureClassDescription(Structure::StructureClass _class)
 }
 
 
+
+static std::array<std::string, StructureID::SID_COUNT> StructureNameTable =
```

Comment:
If StructureNameTable was made a std::map<StructureID, std::string> instead of a std::array, it would facilitate making StructureID an enum class.

I assume std::map is slightly slower than an array lookup, but I'm also guessing the performance here isn't critical enough to matter.

-Brett

Reply:
I'd considered that -- it would also make it easy to add new structure names without having to ensure order is correct. @DanRStevens seemed to suggest otherwise -- I personally prefer the std::map version but std::array is technically faster.

I'm okay with either implementation, really. It's why I opted for an accessor function so that internal implementation can be changed without needing to make changes elsewhere in the code.

Reply:
You're both quite correct in your observations. Using a `std::map` would allow the conversion to be done without using `static_cast`. The performance difference between `std::map` and `std::array` is expected to be quite small, and of no consequence here. The accessor function does hide the `static_cast` in a single method, preventing spread to other code. Assuming the `StructureName` method is used for conversion, the `StructureID` can be upgraded to an `enum class` without adding additional `static_cast` conversions. Effectively higher efficiency can be achieved with low source code cost. I also expect many current lookups can be removed with future refactoring (general code observation, not specific to this enum).

```cpp
std::string StructureName(StructureID id)
{
	return StructureNameTable[static_cast<size_t>(id)];
}
```

----

Btw, why not lowercase for the initial letter in `StructureName`? Other global functions seem to use lowerCameCase.

----

The current version of `StructureName` is not safe for out of range `id` values. Using `.at(index)` instead of `operator[index]` would ensure safety, causing an exception to be raised if an out of bounds `id` value was passed in. Additionally, a manual bounds check can be inserted if a more detailed error message is desired.

```cpp
std::string StructureName(StructureID id)
{
	return StructureNameTable.at(static_cast<size_t>(id));
}
```